### ISSUE TITLE
Add Support for Geth

### ISSUE THREAD

**Issue by zixuanzh** [poanetwork/blockscout#183]
Created a fork to add geth support. This issue can be assigned to me.

**Comment by acravenho** [comment]
@zixuanzh Thank you for your help on this.

**Comment by acravenho** [comment]
See issue #184 for more info on Geth compatibility.

**Comment by peterromfeldhk** [comment]
anything i can do to help? i would like to use this UI for our own POA blockchain running with geth (only sane UI i have seen so far :) ) I tried to change geth.exs URL, but it seems to just use the parity one and goes to mainnet even if i set ETHEREUM_JSONRPC_VARIANT=geth

**Comment by themotu** [comment]
@peterromfeldhk Any luck with this? I change env vars but am only getting odd errors back. What rpc options are you running geth with to make this work?

**Comment by themotu** [comment]
I'm getting errors when I try this with geth or ganache:

```
[error] Task #PID<0.560.0> started from #PID<0.556.0> terminating
** (CaseClauseError) no case clause matching: :ignore
    (indexer) lib/indexer/block_fetcher.ex:137: Indexer.BlockFetcher.insert/2
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Function: &:erlang.apply/2
    Args: [#Function<8.109633770/1 in Indexer.BlockFetcher.stream_import/1>, [0..1]]
[error] Realtime index stream exited with reason ({{:case_clause, :ignore}, [{Indexer.BlockFetcher, :insert, 2, [file: 'lib/indexer/block_fetcher.ex', line: 137]}, {Task.Supervised, :do_apply, 2, [file: 'lib/task/supervised.ex', line: 89]}, {Task.Supervised, :reply, 5, [file: 'lib/task/supervised.ex', line: 38]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}).  The next realtime index task will fill the missing block if the lastest block number has not advanced by then or the catch up index will fill the missing block.
[error] GenServer #PID<0.558.0> terminating
** (CaseClauseError) no case clause matching: :ignore
    (indexer) lib/indexer/block_fetcher.ex:137: Indexer.BlockFetcher.insert/2
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:EXIT, #PID<0.556.0>, {{:case_clause, :ignore}, [{Indexer.BlockFetcher, :insert, 2, [file: 'lib/indexer/block_fetcher.ex', line: 137]}, {Task.Supervised, :do_apply, 2, [file: 'lib/task/supervised.ex', line: 89]}, {Task.Supervised, :reply, 5, [file: 'lib/task/supervised.ex', line: 38]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}}
State: %Indexer.Sequence{current: nil, queue: {[], []}, step: 2}
[error] Task #PID<0.566.0> started from #PID<0.563.0> terminating
** (CaseClauseError) no case clause matching: :ignore
    (indexer) lib/indexer/block_fetcher.ex:137: Indexer.BlockFetcher.insert/2
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Function: &:erlang.apply/2
    Args: [#Function<8.109633770/1 in Indexer.BlockFetcher.stream_import/1>, [0..1]]
[error] Realtime index stream exited with reason ({{:case_clause, :ignore}, [{Indexer.BlockFetcher, :insert, 2, [file: 'lib/indexer/block_fetcher.ex', line: 137]}, {Task.Supervised, :do_apply, 2, [file: 'lib/task/supervised.ex', line: 89]}, {Task.Supervised, :reply, 5, [file: 'lib/task/supervised.ex', line: 38]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}).  The next realtime index task will fill the missing block if the lastest block number has not advanced by then or the catch up index will fill the missing block.
[error] GenServer #PID<0.564.0> terminating
** (CaseClauseError) no case clause matching: :ignore
    (indexer) lib/indexer/block_fetcher.ex:137: Indexer.BlockFetcher.insert/2
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {:EXIT, #PID<0.563.0>, {{:case_clause, :ignore}, [{Indexer.BlockFetcher, :insert, 2, [file: 'lib/indexer/block_fetcher.ex', line: 137]}, {Task.Supervised, :do_apply, 2, [file: 'lib/task/supervised.ex', line: 89]}, {Task.Supervised, :reply, 5, [file: 'lib/task/supervised.ex', line: 38]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}]}}
State: %Indexer.Sequence{current: nil, queue: {[], []}, step: 2}
```

**Comment by peterromfeldhk** [comment]
```
geth --verbosity 1 --syncmode fast --networkid 888 --rpc --rpcaddr
127.0.0.1 --rpcapi web3,eth --rpcport 8545 --rpccorsdomain '*'
```

i can send the whole nixos config for it that would help you

**Comment by peterromfeldhk** [comment]
its still a bit hacky, but no time currently (you will need a `genesis.json` in same dir and also a `static-nodes.json` pointing to your blockchain)

`nixos-modules/geth.nix`
```
{ config, pkgs, lib, ... }:

with lib;

let
  cfg = config.services.my-geth;

  dbDir = "/var/db/geth";
  networkId = "888";
  genesis = ./genesis.json;
  staticNodes = ./static-nodes.json;

  gethPackage = pkgs.go-ethereum;

in {
  options.services.my-geth = {

    enable = mkOption {
      type = types.bool;
      default = false;
    };

    pkJson = mkOption {
      type = types.nullOr types.path;
      default = null;
    };

    syncmode = mkOption {
      type = types.string;
      default = "fast";
    };

    rpcapi = mkOption {
      type = types.string;
      default = "eth,net,web3,personal";
    };

    rpcaddr = mkOption {
      type = types.string;
      default = "127.0.0.1";
    };

    extraArgs = mkOption {
      type = types.string;
      default = "--targetgaslimit 10000000";
    };

  };

  config = mkIf cfg.enable {

    users.extraUsers.geth = {
      description = "Geth daemon user";
      home = dbDir;
      createHome = true;
    };

    systemd.services.geth = let
      activationScript = pkgs.writeScript "activationScript" (''
        #!${pkgs.stdenv.shell} -e

        # Initialize chain
        if test ! -d ${dbDir}/.ethereum/geth; then
          ${gethPackage}/bin/geth init ${genesis}
        fi

        # Writing files to users home directory in nix is a bit messy
        # and geth cannot take a parameter for static nodes

        # Geth changes permission of the file to -w
        # Solve this by removing the file prior to copying
        rm -f ${dbDir}/.ethereum/static-nodes.json
        cp ${staticNodes} ${dbDir}/.ethereum/static-nodes.json

      '' + optionalString (cfg.pkJson != null) ''
        rm -f ${dbDir}/.ethereum/keystore/pk.json
        cp ${cfg.pkJson} ${dbDir}/.ethereum/keystore/pk.json
      '');
    in
    {
      description = "Go ethereum daemon";
      serviceConfig.ExecStartPre = activationScript;
      serviceConfig.ExecStart = "${gethPackage}/bin/geth --verbosity 1 --syncmode ${cfg.syncmode} --networkid 888 --rpc --rpcaddr ${cfg.rpcaddr} --rpcapi ${cfg.rpcapi} ${cfg.extraArgs}";
      serviceConfig.Type = "simple";
      serviceConfig.Restart = "on-failure";
      serviceConfig.PrivateTmp = true;
      serviceConfig.PrivateDevices = true;
      serviceConfig.ProtectHome = true;
      serviceConfig.WorkingDirectory = dbDir;
      serviceConfig.User = "geth";
      wantedBy = [ "multi-user.target" ];
      after = [ "network.target" ];
    };

  };
}
```

`nixos-modules/eth-explorer.nix`
```
{ pkgs, config, lib, ... }:
{
  virtualisation.docker = {
    enable = true;
  };

  systemd.services.eth-explorer = let
    PORT = "4000";
    IMAGE = "enumatech/explorer-test:0.0.8";

    environmentfile = pkgs.writeText "eth-explorer.environmentfile" ''
      MIX_ENV=prod
      ETHEREUM_JSONRPC_VARIANT=EthereumJSONRPC.Geth
      ETHEREUM_JSONRPC_TRANSPORT=EthereumJSONRPC.HTTP
      ETHEREUM_JSONRPC_HTTP=EthereumJSONRPC.HTTP.HTTPoison
      ETHEREUM_JSONRPC_HTTP_URL=http://127.0.0.1:8545
      ECTO_USE_SSL=false
      DATABASE_URL=postgres://dora:dora@127.0.0.1:5432/dora
    '';

  in {
    description = "ETH explorer";
    wantedBy = [ "multi-user.target" ];
    after = [ "geth.service" "postgresql.service" "docker.service" ];
    requires = [ "geth.service" "postgresql.service" "docker.service" ];

    serviceConfig = let
      dockerBin = "${pkgs.docker}/bin/docker";
    in {
      TimeoutStartSec = 0;
      Restart = "always";

      ExecStartPre = [
        "-${dockerBin} stop eth-explorer"
        "-${dockerBin} rm eth-explorer"
      ];
      ExecStart = "${dockerBin} run --name eth-explorer --rm --env-file ${environmentfile} --net=host -p ${PORT}:${PORT} ${IMAGE}";
      ExecStop = "${dockerBin} stop eth-explorer";
    };
  };

}
```

`configuration.nix`
```
{ pkgs, config, lib, ... }:
{
  imports = [
    ../nixos-modules/geth.nix
    ../nixos-modules/eth-explorer.nix
  ];

  services.my-geth.enable = true;
  services.my-geth.syncmode = "fast";
  services.my-geth.rpcapi = "web3,eth";
  services.my-geth.extraArgs = "--rpcport 8545 --rpccorsdomain '*'";

  services.postgresql = {
    enable = true;
    package = pkgs.postgresql100;
    enableTCPIP = true;
    initialScript = pkgs.writeText "backend-initScript" ''
      CREATE USER dora;
      ALTER USER dora PASSWORD 'dora';
      ALTER USER dora WITH SUPERUSER;
      CREATE DATABASE dora;
      CREATE EXTENSION btree_gist;
    '';
  };

  services.nginx = {
    enable = true;

    # Use recommended settings
    recommendedGzipSettings = true;
    recommendedOptimisation = true;
    recommendedProxySettings = true;
    recommendedTlsSettings = true;

    # Only allow PFS-enabled ciphers with AES256
    sslCiphers = "AES256+EECDH:AES256+EDH:!aNULL";

    commonHttpConfig = ''
      # Add HSTS header with preloading to HTTPS requests.
      # Adding this header to HTTP requests is discouraged
      map $scheme $hsts_header {
          https   "max-age=31536000; includeSubdomains; preload";
      }
      add_header Strict-Transport-Security $hsts_header;

      # Enable CSP for your services.
      #add_header Content-Security-Policy "script-src 'self'; object-src 'none'; base-uri 'none';" always;

      # Minimize information leaked to other domains
      add_header 'Referrer-Policy' 'origin-when-cross-origin';

      # Disable embedding as a frame
      add_header X-Frame-Options DENY;

      # Prevent injection of code in other mime types (XSS Attacks)
      add_header X-Content-Type-Options nosniff;

      # Enable XSS protection of the browser.
      # May be unnecessary when CSP is configured properly (see above)
      add_header X-XSS-Protection "1; mode=block";

      # This might create errors
      proxy_cookie_path / "/; secure; HttpOnly; SameSite=strict";
    '';

    # Add any further config to match your needs, e.g.:
    virtualHosts = let
      base = locations: {
      inherit locations;

      forceSSL = true;
      enableACME = true;
    };
    proxy = port: base {
      "/".proxyPass = "http://127.0.0.1:" + toString(port) + "/";
    };
    in {
      # Define example.com as reverse-proxied service on 127.0.0.1:3000
      "explorer.domain.com" = proxy 4000 // { default = true; };
    };
  };

  networking.hostName = "explorer.domain.com";
}

```

**Comment by KronicDeth** [comment]
@bestplay it is low priority as the label says and we now have private Parity nodes for Ethereum Mainnet, Ropsen, and ETC, so using Infura for Ethereum Mainnet, which meant Geth support was required for Ethereum Mainnet is no longer true.  The Infura setup is really only used in the CircleCI tests now.  PRs are of course always welcome.

The simplest answer is that low priority stuff won't happen until all the high and medium stuff is done, but the community can shift priorities.  The general approach is 👍 the description of the issue to vote for the issue.  There's also the option of opening a bounty as Gitcoin has done with some of the other issues to incentivize other community members to make PRs if you can't yourself.

**Comment by acravenho** [comment]
Closed by #1072
